src/rust/engine/fs/src/directory.rs

```diff
@@ -16,17 +18,118 @@ use protos::gen::build::bazel::remote::execution::v2 as remexec;
 
 use crate::PathStat;
 
-type Name = Intern<String>;
+pub const EMPTY_DIRECTORY_DIGEST: DirectoryDigest = DirectoryDigest {
+  digest: EMPTY_DIGEST,
+  tree: None,
+};
 
+/// A Digest for a directory, optionally with its content stored as a DigestTrie.
+///
+/// If a DirectoryDigest has a DigestTrie reference, then its Digest _might not_ be persisted to
+/// the Store. If the DirectoryDigest does not hold a DigestTrie, then that Digest _must_ have been
+/// persisted to the Store (either locally or remotely). The field thus acts likes a cache in some
+/// cases, but in other cases is an indication that the tree must first be persisted (or loaded)
+/// before the Digest may be operated on.
+#[derive(Clone, DeepSizeOf)]
+pub struct DirectoryDigest {
+  pub digest: Digest,
+  pub tree: Option<DigestTrie>,
+}
+
+impl Eq for DirectoryDigest {}
+
+impl PartialEq for DirectoryDigest {
+  fn eq(&self, other: &Self) -> bool {
+    self.digest == other.digest
+  }
+}
+
+impl Debug for DirectoryDigest {
+  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+    let tree = if self.tree.is_some() {
+      "Some(..)"
+    } else {
+      "None"
+    };
+    write!(f, "DirectoryDigest({:?}, tree: {})", self.digest, tree)
+  }
+}
+
+impl DirectoryDigest {
+  /// Creates a DirectoryDigest which asserts that the given Digest represents a Directory structure
+  /// which is persisted in a Store.
+  pub fn new(digest: Digest) -> Self {
+    Self { digest, tree: None }
+  }
+
+  pub fn from_path_stats(
+    path_stats: Vec<PathStat>,
+    file_digests: &HashMap<PathBuf, Digest>,
+  ) -> Result<Self, String> {
+    let path_stats = PathStat::normalize_path_stats(path_stats)?;
```

Comment:
Nit, maybe call this `sorted_path_stats` to make more clear that's what this does.